cilium/CFP-30984-dns-proxy-ha.md

```diff
@@ -0,0 +1,231 @@
+
+This CFP introduces a standalone DNS proxy that can be run alongside cilium agent which should eliminate hard dependency for names that already have policy map entries in place.
+
+## Motivation
+
+Users rely on toFQDN policies to enforce network policies against traffic to remote destinations outside the cluster, typically to blob storage / other services on the internet. Rolling out cilium agent should not result in traffic being dropped. Introducing a high availablility (HA) mode will allow for adoption of toFQDN network policies in critical environments.
+
+## Goals
+
+* Introduce a streaming gRPC API for exchanging FQDN policy related information.
+* Introduce standalone DNS proxy (SDP) that binds on the same port as built-in proxy with SO_REUSEPORT and uses the above mentioned API to notify agent of new DNS resolutions.
+* Leverage the bpf maps to resolve IP address to endpoint ID and identity ID for enforcing L7 DNS policy.
+
+## Non-Goals
+
+* Updating new DNS <> IP mappings when agent is down is out of scope
+
+## Proposal
+
+### Overview
+
+![Standalone DNS proxy Overview](./images/standalone-dns-proxy-overview.png)
+
+There are two parts to enforcing toFQDN network policy. L4 policy enforcement against IP adresses resolved from a FQDN and policy enforcement on DNS requests (L7 DNS policy). In order to enforce L4 policy, per endpoint policy bpf maps need to be updated. We'd like to avoid multiple processes writing entries to policy maps, so the standalone DNS proxy (SDP) needs a mechansim to notify agent of newly resolved FQDN <> IP address mappings. This CFP proposes exposing a new gRPC streaming API from cilium agent to do this. Since the connection is bi-directional, cilium agent can re-use the same connection to notify the SDP of L7 DNS policy changes.
+Additionally SDP also needs to translate IP address to endpoint ID and identity in order to enforce policy by reusing the logic from agent's DNS proxy. Our proposal involves retrieving the endpoint and identity data directly from the `cilium_lxc` and `cilium_ipcache` BPF maps, respectively.
+
+### RPC Methods
+
+Method : UpdateMappings
+
+_rpc UpdatesMappings(steam FQDNMapping) returns (Result){}_
+Request :
+```
+message FQDNMapping {
+    string FQDN = 1;
+    repeated bytes IPS = 2;
+    int32 TTL = 3;
+    bytes client_ip = 4;
+    int32 response_code = 5;
+}
+```
+Response :
+```
+message Result {
+    bool success = 1;
+}
+```
+
+Method : UpdatesDNSRules
+
+_rpc UpdatesDNSRules(stream DNSPolicyRules) returns (Result){}_
+Request :
+```
+message FQDNSelector {
+  string match_name = 1;
+  string match_pattern = 2;
+}
+
+message DNSPolicyRule {
+  string selector_string = 1;
+  repeated FQDNSelector port_rules= 2;
```

Comment:
Yes, we should rename the `FQDNSelector` to `PortRules`. (since they FqdnSelector is used for rules in cilium code base, I believe that is why it was kept like that)
The `selector_string` is the result of the function `String()`(method used to create a string based on `MatchName`/`MatchPattern`) . It is used as a key for the map store in dnsproxy codebase.

Reply:
Brainstorm: Something I'm struggling a little bit with these fields is whether they are the logical parts of rules at a specific level of abstraction, or whether it's taking Cilium internals and converting them into public API. In the latter case the concern I have is that we may end up constraining the way that future Cilium versions work because the API is too tied to the implementation details from today.

Reply:
valid concern. we should name in generic and not tied to cilium internals